src/history/hs.c

```diff
@@ -1382,6 +1415,122 @@ __wt_hs_delete_key_from_ts(
+    wt_timestamp_t hs_ts;
+    uint64_t hs_counter;
+    uint32_t hs_btree_id;
+    int cmp;
+    const char *open_cursor_cfg[] = {WT_CONFIG_BASE(session, WT_SESSION_open_cursor), NULL};
+
+    insert_cursor = NULL;
+    hs_cbt = (WT_CURSOR_BTREE *)hs_cursor;
+    insert_cbt = NULL;
+    WT_CLEAR(hs_key);
+    tombstone = NULL;
+
+    /*
+     * The goal of this fixup function is to move out-of-order content to maintain ordering in the
+     * history store. We do this by removing content with higher timestamps and reinserting it
+     * behind (from search's point of view) the newly inserted update.
+     *
+     * For example, if we're inserting an update at timestamp 3 with value ddd:
+     * btree key ts counter value
+     * 2     foo 5  0       aaa
+     * 2     foo 6  0       bbb
+     * 2     foo 7  0       ccc
+     *
+     * We want to end up with this:
+     * btree key ts counter value
+     * 2     foo 3  0       aaa
+     * 2     foo 3  1       bbb
+     * 2     foo 3  2       ccc
+     * 2     foo 3  3       ddd
+     */
+    for (; ret == 0; ret = hs_cursor->next(hs_cursor)) {
+        /* The normal case will break out of the first loop in one of the conditions below. */
+        WT_ERR(hs_cursor->get_key(hs_cursor, &hs_btree_id, &hs_key, &hs_ts, &hs_counter));
+        if (hs_btree_id != btree->id)
+            break;
+        WT_ERR(__wt_compare(session, NULL, &hs_key, key, &cmp));
+        if (cmp != 0)
+            break;
+
+        /*
+         * If we got here, we've got out-of-order updates in the history store.
+         *
+         * Our strategy to rectify this is to remove all records for the same key with a higher
+         * timestamp than the one that we're inserting on and reinsert them at the same timestamp
+         * that we're inserting with.
+         */
+        WT_ASSERT(session, hs_ts > ts);
+
+        /*
+         * Don't incur the overhead of opening this new cursor unless we need it. In the regular
+         * case, we'll never get here.
+         */
+        if (insert_cursor == NULL) {
+            WT_WITHOUT_DHANDLE(session,
+              ret = __wt_open_cursor(session, WT_HS_URI, NULL, open_cursor_cfg, &insert_cursor));
+            WT_ERR(ret);
+            insert_cbt = (WT_CURSOR_BTREE *)insert_cursor;
+            F_SET(insert_cbt, WT_CBT_NO_TXN);
+        }
+
+        start_time_point.ts = start_time_point.durable_ts = ts;
+        start_time_point.txnid = hs_cbt->upd_value->tw.start_txn;
+
+        /*
+         * We're going to be inserting something immediately after with the same timestamp. Either
+         * another moved update OR the update itself that triggered the correction.
+         */
+        stop_time_point.ts = stop_time_point.durable_ts = ts;
+
+        /*
+         * This means it's the latest record for that key. We should give it a stop pair to
+         * correspond to the update we're inserting.
+         */
+        if (hs_cbt->upd_value->tw.stop_txn == WT_TXN_MAX)
```

Comment:
I didn't clearly understand this case. Only for prepare case, we have an update with stop timestamp/transaction with MAX. But this should be fixed when the prepared update is resolved. So I shouldn't expect this case.

Reply:
Yeah, I misunderstood. I think I can just remove this condition.

Reply:
0e9ebd0